### PR TITLE
feat: add new arch

### PR DESCRIPTION
**Pull request by WoLewicki**
## PR adding New Architecture support to the library :tada:

We at [Software Mansion](https://swmansion.com/) have been working on [improving support](https://blog.swmansion.com/sunrising-new-architecture-in-the-new-expensify-app-729d237a02f5) for the new architecture for quite a while now. If you need help with anything related to New Architecture, like:
- [migrating your library](https://x.com/swmansion/status/1717512089323864275)
- [migrating your app](https://github.com/Expensify/App/pull/13767)
- [investigating issues](https://github.com/facebook/react-native/pulls?q=sort%3Aupdated-desc+is%3Apr+author%3Aj-piasecki+is%3Aopen)
- [improving performance](https://x.com/BBloniarz_/status/1808138585528303977)

or you just want to ask any questions, hit us up on [projects@swmansion.com](mailto:projects@swmansion.com)

---

### CODE REVIEW

**Comment by zoontek** [comment]
Hi @WoLewicki 👋 and thanks for the pull request.

Could you drop the `FabricExample` directory? It adds a lot of noise and I'd rather write a `README.md` inside the current example project explaining what's needed to be updated to enable the new arch 🙂

**Comment by WoLewicki** [comment]
Ok, I'll do it when the code will work just fine on both archs 😅

**Comment by WoLewicki** [comment]
@zoontek I believe the PR is ready for testing 🚀

**Comment by zoontek** [comment]
@WoLewicki Could you rebase against master? No worry, I took care of your podspec changes when shipping v3.7.0, you can resolve the conflict using you version.

(I also updated the example to react native 0.71, so it will be easier to review the PR)

**Comment by WoLewicki** [comment]
@zoontek I did it. Looks like you can run new arch build on `example` project now 🚀

**Comment by zoontek** [comment]
@WoLewicki Everything looks good to me, thanks a lot!
I'm merging this, will plan a release during the week 🎉

File: android/build.gradle [hunk 1]
```diff
@@ -19,7 +19,57 @@ buildscript {
...
+    if (reactNativeFromNodeModulesWithLibrary.exists()) {
+        return reactNativeFromNodeModulesWithLibrary
+    }
+
+    throw new Exception(
+            "[react-native-clipboard] Unable to resolve react-native location in " +
+                    "node_modules. You should add project extension property (in app/build.gradle) " +
+                    "`REACT_NATIVE_NODE_MODULES_DIR` with path to react-native."
+    )
+}
```
Comment (trippingtarballs) [inline android/build.gradle:40]: I am working with a monorepo using [NPM workspaces](https://docs.npmjs.com/cli/v9/using-npm/workspaces?v=true) and I can confirm that this implementation of `resolveReactNativeDirectory()` does not enable monorepo support.

## Problem

The values for both `reactNativeFromProjectNodeModules` and `reactNativeFromNodeModulesWithLibrary` are the same!

In a monorepo setup there is likely to be a `node_modules` local to the package's folder, **and** a `node_modules` at the root of the monorepo. I suspect that the intention of having both a `reactNativeFromProjectNodeModules` and `reactNativeFromNodeModulesWithLibrary` was for one to point to each of the two possible locations where the `react-native` library might be located.

## Setup

My workspaces is configured as such in the root `package.json`:

```json
     "workspaces": [
       "apps/*"
     ]
```

I have a folder structure like this:

```
.
├── apps
│   └── rn-app
│       ├── android
│       ├── ios
│       ├── node_modules
│       │   └── react-native-permissions
│       ├── src
│       ├── index.js
│       ├── package.json
│       ├── react-native.config.js
│       └── tsconfig.json
├── node_modules
│   ├── react-native
│   │   └── ...
│   └── ...
├── README.md
├── package-lock.json
└── package.json
```

And, in the file, `apps/rn-app/package.json` I have a reference to `react-native-permissions@3.8.4` under the `dependencies`. 

## Cause

Due to how NPM is resolving my dependencies my reference to `react-native-permissions` has not been hoisted to the `node_modules` at the root of my project (although `react-native` has been hoisted). Consequently, the exception message is triggering. Alas, I have no idea what the instruction of the exception message is asking me to do.

Comment (SerSerch) [reply android/build.gradle:40]: @trippingtarballs Hello!
I also have a monorepo and everything works for me. take a look at my implementation.
https://github.com/serserm/react-native-turbo-sensors

modified files:
[react-native.config.js](https://github.com/serserm/react-native-turbo-sensors/blob/main/example/react-native.config.js)
[metro.config.js](https://github.com/serserm/react-native-turbo-sensors/blob/main/example/metro.config.js)
[babel.config.js](https://github.com/serserm/react-native-turbo-sensors/blob/main/example/babel.config.js)
[settings.gradle](https://github.com/serserm/react-native-turbo-sensors/blob/main/example/android/settings.gradle)
[settings.gradle](https://github.com/serserm/react-native-turbo-sensors/blob/main/packages/react-native-turbo-sensors/android/settings.gradle)
[tsconfig.json](https://github.com/serserm/react-native-turbo-sensors/blob/main/packages/react-native-turbo-sensors/tsconfig.json)

**Pull request merged**